zebra-state/src/config.rs

```diff
@@ -123,48 +127,105 @@ impl Default for Config {
     }
 }
 
-/// Check if there are old database folders and delete them from the filesystem.
+// Cleaning up old database versions
+
+/// Spawns a task that checks if there are old database folders,
+/// and deletes them from the filesystem.
 ///
 /// Iterate over the files and directories in the databases folder and delete if:
 /// - The state directory exists.
 /// - The entry is a directory.
-/// - The directory name has a length of at least 2 characters.
 /// - The directory name has a prefix `v`.
 /// - The directory name without the prefix can be parsed as an unsigned number.
 /// - The parsed number is lower than the hardcoded `DATABASE_FORMAT_VERSION`.
-pub async fn check_and_delete_old_databases(config: Config) {
+pub fn check_and_delete_old_databases(config: Config) -> JoinHandle<()> {
+    let current_span = Span::current();
+
+    spawn_blocking(move || {
+        current_span.in_scope(|| {
+            delete_old_databases(config);
+            info!("finished old database version cleanup task");
+        })
+    })
+}
+
+/// Check if there are old database folders and delete them from the filesystem.
+///
+/// See [`check_and_delete_old_databases`] for details.
+fn delete_old_databases(config: Config) {
     if config.ephemeral || !config.delete_old_database {
         return;
     }
 
-    info!("checking old database versions");
-    let cache_dir = config.cache_dir.join("state");
-    if let Some(read_dir) = read_dir(cache_dir.clone()) {
-        for entry in read_dir.flatten() {
-            if let Some(dir_name) = parse_dir_name(entry) {
-                if let Some(version_number) = parse_version_number(dir_name.clone()) {
-                    if version_number < crate::constants::DATABASE_FORMAT_VERSION {
-                        let delete_path = cache_dir.join(dir_name);
-                        if remove_dir_all(delete_path.clone()).is_ok() {
-                            info!("deleted outdated state directory {:?}", delete_path);
-                        }
-                    }
-                }
+    info!("checking for old database versions");
+
+    let state_dir = config.cache_dir.join("state");
+    if let Some(state_dir) = read_dir(&state_dir) {
```

Comment:
I was a bit confused by the different paths both named cache_dir, so I renamed one of them.

Reply:
Yea, the join was in the start.rs file in a previous version but had to move it when spanwing in a new task. Forgot to rename one of those variables.